compiler/src/iree/compiler/Reducer/Framework/Delta.cpp

```diff
@@ -0,0 +1,124 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Reducer/Framework/Delta.h"
+
+using namespace mlir;
+using namespace mlir::iree_compiler;
+
+FailureOr<WorkItem> Delta::checkChunk(Chunk maybeUninterestingChunk,
+                                      DeltaFunc deltaFunc,
+                                      ArrayRef<Chunk> maybeInterestingChunks,
+                                      DenseSet<Chunk> &uninterestingChunks) {
+  SmallVector<Chunk> currentChunks;
+  copy_if(maybeInterestingChunks, std::back_inserter(currentChunks),
+          [&](Chunk chunk) {
+            return chunk != maybeUninterestingChunk &&
+                   !uninterestingChunks.count(chunk);
+          });
+
+  debugOs << "Checking chunk: \n";
+  maybeUninterestingChunk.dump();
+  debugOs << "Saved chunks: \n";
+  for (Chunk c : currentChunks) {
+    c.dump();
+  }
+
+  ChunkManager chunker(currentChunks);
+  WorkItem clonedProgram = root.clone();
+  deltaFunc(chunker, clonedProgram);
+
+  if (!oracle.isInteresting(clonedProgram)) {
+    debugOs << "Chunk is uninteresting\n";
+    clonedProgram.getModule()->erase();
+    return failure();
+  }
+
+  debugOs << "Chunk is interesting\n";
+  return clonedProgram;
+};
+
+bool Delta::increaseGranuality(SmallVector<Chunk> &chunks) {
+  debugOs << "Increasing granularity\n";
+  SmallVector<Chunk> newChunks;
+  bool anyNewSplit = false;
+
+  for (Chunk c : chunks) {
+    if (c.getEnd() - c.getBegin() <= 1) {
+      newChunks.push_back(c);
+    } else {
+      int half = (c.getEnd() + c.getBegin()) / 2;
+      newChunks.push_back(Chunk(c.getBegin(), half));
+      newChunks.push_back(Chunk(half, c.getEnd()));
+      anyNewSplit = true;
+    }
+  }
+
+  if (anyNewSplit) {
+    chunks = newChunks;
+    debugOs << "Successfully increased granularity\n";
+    debugOs << "New Chunks:\n";
+    for (Chunk c : chunks) {
+      c.dump();
```

Comment:
dumping is very expensive and should be compiled out (LLVM_DEBUG/etc) - it also goes to llvm::dbgs() which may be different than debugOs (which is why it should either be llvm::dbgs() instead of passing in a different stream _or_ use print() that takes a stream)